tensorflow_model_optimization/python/core/sparsity_tf2/sparse_utils_test.py

```diff
@@ -0,0 +1,223 @@
+    type_perm_checks.append(sparse_permuteones.dtype)
+    self.assertAllEqual(_dtype, type_perm_checks)
+    
+  @parameterized.parameters(
+    itertools.product((0.4,), NOSMALL_SHAPES)
+  )
+  def testScatteredInitialOnesPermuteOnes(self, ratio, shape):
+    # This test checks that the expected sparsity is reflected in the 
+    # average number of times each index in the matrix is a 1.
+    # note that ratio here is fixed because the shape / density
+    # would make a greater difference in this check for "scattered-ness".
+    # Shapes cannot be small (i.e. () or (1) ) as there are not enough
+    # samples to average over.
+    matrix = tf.ones(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    seed = 0
+    sparse_matrices_perm = []
+    for _ in range(400):
+      matrix_sparse_permute = sparse_utils.PermuteOnes(ratio)(shape, seed=seed)
+      sparse_matrices_perm.append(matrix_sparse_permute)
+      seed += 1
+    average_initialization = tf.math.reduce_mean(sparse_matrices_perm, axis=0)
+    expected_initialization = tf.constant(ratio, shape=shape) # broadcasts
+    # expect that each entry in the matrix is "on"/1 in expectation, ratio number of times
+    self.assertAllClose(expected_initialization, average_initialization, rtol=231e-3)
+   
+  @parameterized.parameters(
+    itertools.product((0.4,), NOSMALL_SHAPES)
+  )
+  def testScatteredInitialOnesBernouilli(self, ratio, shape):
+    matrix = tf.ones(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    seed = 0
+    sparse_matrices_bern = []
+    for _ in range(400):
+      matrix_sparse_bernouilli = sparse_utils.Bernouilli(ratio)(shape, seed=seed)
+      sparse_matrices_bern.append(matrix_sparse_bernouilli)
+      seed += 1
+    average_initialization = tf.math.reduce_mean(sparse_matrices_bern, axis=0)
+    expected_initialization = tf.constant(ratio, shape=shape)
+    self.assertAllClose(expected_initialization, average_initialization, rtol=204e-3)
+
+  @parameterized.parameters(
+    zip(itertools.product((0.4,), NOSMALL_SHAPES), MIDPOINTS)
+  )
+  def testMidpointInitialOnesPermuteOnes(self, ratio_shape_pair, midpoint):
+    # Given all indices that are masked according to ratio,
+    # the average value of all indices should appear at the "center"
+    # of the given matrix. Similarly to above test, the ratio
+    # of 1's in the mask is not deciding factor.
+    ratio, shape = ratio_shape_pair
+    matrix = tf.ones(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    seed = 0
+    sparse_matrices_perm = []
+    for _ in range(400):
+      matrix_sparse_permute = sparse_utils.PermuteOnes(ratio)(shape, seed=seed)
+      matrix_midpoint_idx = tf.where(tf.reshape(matrix_sparse_permute, (-1,)) == 1)
+      sparse_matrices_perm.append(tf.math.reduce_mean(matrix_midpoint_idx))
+      seed += 1
+    average_midpoints = tf.math.reduce_mean(sparse_matrices_perm, axis=0)
+    expected_midpoints = tf.constant(midpoint) # broadcasts
```

Comment:
Some of the shapes aren't symmetric but you have a single `midpoint` scalar for each shape, how is your check working there? Is it checking just one dimension, or is it checking against the wrong value in some of the dimensions?

(You may be able to reduce the atol if you fix this)

Reply:
I am basically checking against a "flattened" version of  the shapes, because I'm appending the indices of where the masks are 1's and comparing the mean value of those indices (which I expect to be the mean index of the entire flattened matrix)

Reply:
Can you add a comment that explains this & points out the reshape/flattening?